commit 7459a3694730463b8470961ab0f8ebdb62c68cca
Author: Konstantin Kolinko <kkolinko@apache.org>
Date:   Thu Nov 3 14:40:21 2011 +0000

    Add new attribute to AbstractEndpoint and use it to speed up Tomcat tests.
    If the attribute "fastShutdown" is set on the endpoint,
    the usual wait of 1 sec during pause() is skipped.
    
    I think it needs to be added to http.xml/ajp.xml docs, but that will be a separate commit sometime later.
    
    git-svn-id: https://svn.apache.org/repos/asf/tomcat/trunk@1197158 13f79535-47bb-0310-9956-ffa450edef68

diff --git a/java/org/apache/tomcat/util/net/AbstractEndpoint.java b/java/org/apache/tomcat/util/net/AbstractEndpoint.java
index 0038589..a90a29f 100644
--- a/java/org/apache/tomcat/util/net/AbstractEndpoint.java
+++ b/java/org/apache/tomcat/util/net/AbstractEndpoint.java
@@ -112,6 +112,23 @@ public abstract class AbstractEndpoint {
 
     // ----------------------------------------------------------------- Properties
 
+    /**
+     * A flag that can be used to speed up Tomcat shutdown by testing
+     * environments where we control external connections to Tomcat. Set it to
+     * {@code true} if it is known that there are no active or pending
+     * connections and all requests have already been processed. The default
+     * value is {@code false}.
+     */
+    private boolean fastShutdown = false;
+
+    public void setFastShutdown(boolean fastShutdown) {
+        this.fastShutdown = fastShutdown;
+    }
+
+    public boolean isFastShutdown() {
+        return fastShutdown;
+    }
+
     private int maxConnections = 10000;
     public void setMaxConnections(int maxCon) {
         this.maxConnections = maxCon;
@@ -514,6 +531,11 @@ public abstract class AbstractEndpoint {
     public void pause() {
         if (running && !paused) {
             paused = true;
+            if (isFastShutdown()) {
+                // unlockAccept will also be called by stopInternal(),
+                // so when shutting down it can be skipped here.
+                return;
+            }
             unlockAccept();
             // Heuristic: Sleep for a while to ensure pause of the endpoint
             try {
diff --git a/test/org/apache/catalina/startup/TomcatBaseTest.java b/test/org/apache/catalina/startup/TomcatBaseTest.java
index 619a777..4f3f3d1 100644
--- a/test/org/apache/catalina/startup/TomcatBaseTest.java
+++ b/test/org/apache/catalina/startup/TomcatBaseTest.java
@@ -200,6 +200,10 @@ public abstract class TomcatBaseTest {
 
     @After
     public void tearDown() throws Exception {
+        // Speed up Tomcat shutdown
+        if (tomcat.connector != null) {
+            tomcat.connector.setProperty("fastShutdown", "true");
+        }
         // Some tests may call tomcat.destroy(), some tests may just call
         // tomcat.stop(), some not call either method. Make sure that stop() &
         // destroy() are called as necessary.
